pkg/openshift/rosa/rosa.go

```diff
@@ -0,0 +1,225 @@
+
+	gzipReader, err := gzip.NewReader(tarFileReader)
+	if err != nil {
+		return "", fmt.Errorf("failed to create gzip reader for %s: %v", tarFilename, err)
+	}
+	defer gzipReader.Close()
+
+	tarReader := tar.NewReader(gzipReader)
+
+	for {
+		_, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			break
+		}
+		_, err = io.Copy(rosaFile, tarReader)
+		if err != nil {
+			break
+		}
+	}
+
+	return rosaFilename, nil
+}
+
+// versionCheck verifies the rosa cli version meets the minimal version required
+func versionCheck(ctx context.Context, rosaBinary string) error {
+	stdout, _, err := cmd.Run(exec.CommandContext(ctx, rosaBinary, "version"))
+	if err != nil {
+		return err
+	}
+
+	versionSlice := strings.SplitAfter(fmt.Sprint(stdout), "\n")
+	if len(versionSlice) == 0 {
+		return fmt.Errorf("versionCheck failed to get version from cli standard out")
+	}
+
+	currentVersion, err := semver.NewVersion(strings.ReplaceAll(versionSlice[0], "\n", ""))
+	if err != nil {
+		return fmt.Errorf("versionCheck failed to parse version to semantic version: %v", err)
+	}
+
+	minVersion, err := semver.NewVersion(minimumVersion)
+	if err != nil {
+		return fmt.Errorf("versionCheck failed to parse minimum version to semantic version: %v", err)
+	}
+
+	if minVersion.Compare(currentVersion) == 1 {
+		return fmt.Errorf("current rosa version is %q and must be >= %q", currentVersion.String(), minVersion)
+	}
+
+	return nil
+}
+
+// verifyCredentials validates the ocm token and aws credentials to ensure they are valid
+func verifyCredentials(ctx context.Context, rosaBinary string, token, environment string, awsCredentials *awscloud.AWSCredentials) error {
+	commandArgs := []string{"login", "--token", token, "--env", environment}
+
+	command := exec.CommandContext(ctx, rosaBinary, commandArgs...)
+	command.Env = append(command.Environ(), awsCredentials.CredentialsAsList()...)
+
+	_, _, err := cmd.Run(command)
+	if err != nil {
+		return fmt.Errorf("login failed %v", err)
+	}
+	return nil
+}
+
+// New handles constructing the rosa provider which creates a connection
+// to openshift cluster manager "ocm". It is the callers responsibility
+// to close the ocm connection when they are finished (defer provider.Connection.Close())
+func New(ctx context.Context, token string, environment ocmclient.Environment, args ...any) (*Provider, error) {
+	if environment == "" || token == "" {
+		return nil, &providerError{err: fmt.Errorf("some parameters are undefined, unable to construct osd provider")}
+	}
+
+	rosaBinary, err := cliCheck()
+	if err != nil {
+		return nil, &providerError{err: err}
+	}
+
+	err = versionCheck(ctx, rosaBinary)
+	if err != nil {
+		return nil, &providerError{err: err}
+	}
+
+	awsCredentials := &awscloud.AWSCredentials{}
+	if len(args) == 1 {
+		awsCredentials = args[0].(*awscloud.AWSCredentials)
+	} else if len(args) > 1 {
+		return nil, &providerError{err: fmt.Errorf("only one AWSCredentials can be provided")}
+	}
```

Comment:
is this the only type that args can be? If so can we make it `...*awscloud.AWSCredentials`? That would allow you to only do the length check

Reply:
As of now, yes. Just this one type. Resolved in aa56d8e78a0aaee0b0ce2c202e8a1855f0ab7ace